Add DS custom environments (#258)
seqeralabs/docs: platform_versioned_docs/version-24.1/data_studios/custom-envs.mdx

---
title: "Custom environments"
description: "Custom environments for Data Studios"
date: "1 Oct 2024"
tags: [data, studios]
---

In addition to the Seqera-provided container template images, you can provide your own custom container environments, either by augmenting the Seqera-provided images with a list of Conda packages that you provide or by providing your own base container template image.

Data Studios uses the [Wave][wave-home] service to build custom container template images.

## Conda packages

### Prerequisites

- Wave must be configured. For more information, see [Wave containers][wave].
 
 :::note

>  Enterprise deployments must either allow access to the public Wave server, or self-host their own Wave server.

Reviewer: ```suggestion
 To augment Seqera-provided images, Enterprise deployments must either allow access to the public Wave server, or self-host their own Wave server.
```

 :::

### Conda package syntax {#conda-package-syntax}

When adding a new data studio session, you can customize its configuration to install a list of Conda packages to the template image. The supported schema is identical to that used by the Conda `environment.yml` environment file. For more information on the Conda environment file, see [Creating an environment file manually][env-manually].

```yaml title="Example environment.yml file"
name: example-env
channels:
  - defaults
  - conda-forge
dependencies:
  - python=3.7
  - numpy
  - pandas
  - scikit-learn
  - pip:
    - matplotlib
    - seaborn
    - jupyter
```

To create a data studio with custom Conda packages, see [Add a data studio][add-ds].

## Custom container template image {#custom-containers}

For advanced use cases, you can build your own container template image.

:::note

> Amazon Elastic Container Registry (ECR) is the only supported private container registry.

Reviewer: ```suggestion
Public container registries are supported by default. Amazon Elastic Container Registry (ECR) is the only supported private container registry.
```

To create a data studio with a custom template image, see [Add a data studio][add-ds].

#### Ports

The container must use the value of the `CONNECT_TOOL_PORT` environment variable as the listening port for whatever interactive software that you include in your custom container.

#### Signals

Upon termination, the container's main process must handle the `SIGTERM` signal and perform any necessarily cleanup. After a 30 second grace period, the container receives the `SIGKILL` signal.

#### Minimal Dockerfile

The minimal Dockerfile includes directives to accomplish the following:

- Pulls a Seqera-provided base image with prerequisite binaries
- Installs and configures Seqera Fusion
- Sets the container entry point

Customize the following Dockerfile to include any additional software that you require:

```docker title="Minimal Dockerfile"
# Seqera base image
FROM public.cr.seqera.io/platform/connect-client:0.7 as connect

# 1. Add connect binary
COPY --from=connect /connect-client /usr/bin/connect-client

# 2. Install connect dependencies
RUN /usr/bin/connect-client --install

# 3. Configure connect as the entrypoint

> ENTRYPOINT ["/usr/bin/connect-client"]

Reviewer 2: with the latest changes this line needs to be `ENTRYPOINT ["/usr/bin/connect-client", "--entrypoint"]`

```

> For example, to run a basic Python-based HTTP server, build a container from the following Dockerfile. When a data studio runs the custom template environment, the value for the `CONNECT_TOOL_PORT` environment variable is provided dynamically.

Reviewer: @jason-seqera Have you tested that this actually works?

```docker title="Example Dockerfile with Python HTTP server"
FROM public.cr.seqera.io/platform/connect-client:0.7 as connect

FROM ubuntu:20.04
RUN apt-get update --yes && apt-get install --yes --no-install-recommends python3

COPY --from=connect /connect-client /usr/bin/connect-client
RUN /usr/bin/connect-client --install

> ENTRYPOINT ["/usr/bin/connect-client"]

Reviewer 2: with the latest changes this line needs to be `ENTRYPOINT ["/usr/bin/connect-client", "--entrypoint"]`

CMD ["/usr/bin/bash", "-c", "python3 -m http.server $CONNECT_TOOL_PORT"]
```

### Inspect container augmentation build status {#build-status}

You can inspect the progress of a custom container template image build, including any errors if the build fails. A link to the [Wave service][wave-home] container build report is always available for builds.

If the custom container template image build fails, the data studio session has the **build-failed** status. The details about build failures are available when inspecting the session details in the **Error report** tab.

To inspect the status of an ongoing build, or a successful or failed build, complete the following steps:

1. Select the **Data Studios** tab in Seqera Platform.

> 1. From the list of data studio sessions, select the name of the session with **building** or **build-failed** status that you want to inspect.

Reviewer: ```suggestion
1. From the list of data studio sessions, select the name of the session with **building** or **build-failed** status that you want to inspect and click the "View" button.
```

---
title: "Custom environments"
description: "Custom environments for Data Studios"
date: "1 Oct 2024"
tags: [data, studios]
---

In addition to the Seqera-provided container template images, you can provide your own custom container environments, either by augmenting the Seqera-provided images with a list of Conda packages that you provide or by providing your own base container template image.

## Conda packages

### Prerequisites

> - Wave must be configured. For more information, see [Wave containers][wave].

Reviewer: ```suggestion
- Wave must be configured. For more information, see [Wave containers][wave].
 
 :::note
 Enterprise deployments must either allow access to the public Wave server, or self-host their own Wave server.
 :::
```

### Conda package syntax {#conda-package-syntax}

> When you add a new data studio, you can customize its configuration to install a list of Conda packages to the template. The supported schema is identical to that used by the Conda `environment.yml` environment file. For more information on the Conda environment file, see [Creating an environment file manually][env-manually].

Reviewer: ```suggestion
When adding a new data studio session, you can customize its configuration to install a list of Conda packages to the template image. The supported schema is identical to that used by the Conda `environment.yml` environment file. For more information on the Conda environment file, see [Creating an environment file manually][env-manually].
```

```yaml title="Example environment.yml file"
name: example-env
channels:
  - defaults
  - conda-forge
dependencies:
  - python=3.7
  - numpy
  - pandas
  - scikit-learn
  - pip:
    - matplotlib
    - seaborn
    - jupyter
```

To create a data studio with custom Conda packages, see [Add a data studio][add-ds].

## Custom container template image {#custom-containers}

For advanced use cases, you can build your own container template image.

:::note

> Amazon Elastic Container Registry (ECR) is the only supported private container registry.

Reviewer: ```suggestion
Amazon Elastic Container Registry (ECR) is currently the only supported private container registry.
```

Author: @robnewman, per Google Developer Style, use of temporal terms is discouraged.

> currently [link](https://developers.google.com/style/word-list#currently)
> Avoid because this word is implied. The word can also prematurely disclose product or feature strategy or inappropriately imply that a product or feature might change.

So that's why I omitted it here from the original version that had it.

Reviewer: Okay - please omit.

:::

### Prerequisites

> - Access to a container image repository, either a public container registry or in a private Amazon ECR repository

Reviewer: I'm not sure if users need to have credentials added to the workspace in order to access the container registry

```suggestion
- Access to a container image repository, either a public container registry or a private Amazon ECR repository
```

- A container template image

### Dockerfile configuration {#dockerfile}

> For your custom template container image, you must use a Seqera provided base image and include several additional build steps for compatibility with Data Studios.

Reviewer: ```suggestion
For your custom template container image, you must use a Seqera-provided base image and include several additional build steps for compatibility with Data Studios.
```

#### Ports

The container must use the value of the `CONNECT_TOOL_PORT` environment variable as the listening port for whatever interactive software that you include in your custom container.

#### Signals

Upon termination, the container's main process must handle the `SIGTERM` signal and perform any necessarily cleanup. After a 30 second grace period, the container receives the `SIGKILL` signal.

#### Minimal Dockerfile

The minimal Dockerfile includes directives to accomplish the following:

- Pulls a Seqera-provided base image with prerequisite binaries

> - Installs prerequisite packages, such as wget and fuse

Reviewer: ```suggestion
- Installs prerequisite packages, such as `wget` and `fuse`
```

Author: @robnewman, this is always a bit confounding, as for example the package name that is installed on say Debian is actually called _wget_, and the binary itself is called `wget`. Looking at the Dockerfile, we install these packages, which then provide the binaries of the same name.

This is similar to in K8s, where we might have a _pod_ that is configured with the Pod resource. Is it Pod or pod? The distinction isn't always clear. In this case, I think just _wget_ or _fuse_ is closer to the mark.

Reviewer: Okay - sounds good. Ignore my suggestion.

- Installs and configures Seqera Fusion
- Sets the container entry points

Customize the following Dockerfile to include any additional software that you require:

```docker title="Minimal Dockerfile"
# Seqera base image
FROM public.cr.seqera.io/platform/connect-client:0.7.1 as connect

# 1. Add connect binary
COPY --from=connect /connect-client /usr/bin/connect-client

# 2. Install required system packages
RUN apt-get update --yes && apt-get install --yes --no-install-recommends \
    ca-certificates \
    wget \
    fuse \
    btrfs-progs \
    squashfs-tools \
    && apt-get clean && rm -rf /var/lib/apt/lists/*

> # 3. Install Fusion

Reviewer 3: Really dumb question: why don't we just provide a base image with Fusion pre-baked, the connect client in the right place etc, so users don't have to do all this every time?

Reviewer: I think this is the only outstanding issue to be resolved.

RUN wget https://fusionfs.seqera.io/releases/pkg/2/4/2/fusion-amd64.tar.gz && \
    tar xvzf fusion-amd64.tar.gz && \
    rm fusion-amd64.tar.gz

# 4. Configure Fusion env variables
ENV FUSION_LOG_OUTPUT=/tmp/.fusion.log
ENV FUSION_COMMAND_BUFFER_SIZE=0
ENV FUSION_VALIDATE_WORK=false

# 5. Configure connect and Fusion as the entrypoint
ENTRYPOINT ["/usr/bin/fusion", "/usr/bin/connect-client"]
```

> To run a basic Python-based HTTP server, build a container from the following Dockerfile. When a data studio runs the custom template environment, the value for the `CONNECT_TOOL_PORT` environment variable is provided dynamically.

Reviewer 3: ```suggestion
For example, to run a basic Python-based HTTP server, build a container from the following Dockerfile. When a data studio runs the custom template environment, the value for the `CONNECT_TOOL_PORT` environment variable is provided dynamically.
```

```docker title="Example Dockerfile with Python HTTP server"

> FROM public.cr.seqera.io/platform/connect-client:0.7.1 as connect

Reviewer 4: The latest image should be: `cr.seqera.io/public/connect-client:0.7.4`, but we may want to allow some wiggle for security and minor updates, so unless there are objections @robnewman, we should use:

`cr.seqera.io/public/connect-client:0.7`

Reviewer: Sounds good - agreed!

FROM ubuntu:20.04
RUN apt-get update --yes && apt-get install --yes --no-install-recommends python3

COPY --from=connect /connect-client /usr/bin/connect-client

RUN apt-get update --yes && apt-get install --yes --no-install-recommends \
    ca-certificates \
    wget \
    fuse \
    btrfs-progs \
    squashfs-tools \
    && apt-get clean && rm -rf /var/lib/apt/lists/*

RUN wget https://fusionfs.seqera.io/releases/pkg/2/4/2/fusion-amd64.tar.gz && \
    tar xvzf fusion-amd64.tar.gz && \
    rm fusion-amd64.tar.gz

ENV FUSION_LOG_OUTPUT=/tmp/.fusion.log
ENV FUSION_COMMAND_BUFFER_SIZE=0
ENV FUSION_VALIDATE_WORK=false

ENTRYPOINT ["/usr/bin/fusion", "/usr/bin/connect-client"]

CMD ["/usr/bin/bash", "-c", "python3 -m http.server $CONNECT_TOOL_PORT"]

```

To create a data studio with a custom template image, see [Add a data studio][add-ds].

> 

Reviewer: Need to add a section about what happens if you use a custom environment (i.e. that Wave is used to build the container and that the list-view UI and interstitial will have the status `building`). Also, what happens if the `build-failed` and what to do about it (i.e. view error logs).

```docker title="Minimal Dockerfile"
# Seqera base image
FROM public.cr.seqera.io/platform/connect-client:0.7.1 as connect

# 1. Add connect binary
COPY --from=connect /connect-client /usr/bin/connect-client

# 2. Install required system packages
RUN apt-get update --yes && apt-get install --yes --no-install-recommends \
    ca-certificates \
    wget \
    fuse \
    btrfs-progs \
    squashfs-tools \
    && apt-get clean && rm -rf /var/lib/apt/lists/*

# 3. Install Fusion
RUN wget https://fusionfs.seqera.io/releases/pkg/2/4/2/fusion-amd64.tar.gz && \
    tar xvzf fusion-amd64.tar.gz && \
    rm fusion-amd64.tar.gz

# 4. Configure Fusion env variables
ENV FUSION_LOG_OUTPUT=/tmp/.fusion.log
ENV FUSION_COMMAND_BUFFER_SIZE=0
ENV FUSION_VALIDATE_WORK=false

# 5. Configure connect and Fusion as the entrypoint
ENTRYPOINT ["/usr/bin/fusion", "/usr/bin/connect-client"]
```

For example, to run a basic Python-based HTTP server, build a container from the following Dockerfile. When a data studio runs the custom template environment, the value for the `CONNECT_TOOL_PORT` environment variable is provided dynamically.

```docker title="Example Dockerfile with Python HTTP server"

> FROM public.cr.seqera.io/platform/connect-client:0.7.1 as connect

Reviewer: ```suggestion
FROM public.cr.seqera.io/platform/connect-client:0.7 as connect
```

#### Minimal Dockerfile

The minimal Dockerfile includes directives to accomplish the following:

- Pulls a Seqera-provided base image with prerequisite binaries
- Installs prerequisite packages, such as wget and fuse
- Installs and configures Seqera Fusion
- Sets the container entry points

Customize the following Dockerfile to include any additional software that you require:

```docker title="Minimal Dockerfile"
# Seqera base image
FROM public.cr.seqera.io/platform/connect-client:0.7 as connect

# 1. Add connect binary
COPY --from=connect /connect-client /usr/bin/connect-client

# 2. Install required system packages
RUN apt-get update --yes && apt-get install --yes --no-install-recommends \
    ca-certificates \
    wget \
    fuse \
    btrfs-progs \
    squashfs-tools \
    && apt-get clean && rm -rf /var/lib/apt/lists/*

# 3. Install Fusion
RUN wget https://fusionfs.seqera.io/releases/pkg/2/4/2/fusion-amd64.tar.gz && \
    tar xvzf fusion-amd64.tar.gz && \
    rm fusion-amd64.tar.gz

# 4. Configure Fusion env variables
ENV FUSION_LOG_OUTPUT=/tmp/.fusion.log
ENV FUSION_COMMAND_BUFFER_SIZE=0
ENV FUSION_VALIDATE_WORK=false

# 5. Configure connect and Fusion as the entrypoint

> ENTRYPOINT ["/usr/bin/fusion", "/usr/bin/connect-client"]

Reviewer 2: Updated installation guide:
```
# Seqera base image
FROM public.cr.seqera.io/platform/connect-client:0.7 as connect

# 1. Add connect binary
COPY --from=connect /connect-client /usr/bin/connect-client

# 2. Install connect dependencies
RUN /usr/bin/connect-client --install

# 3. Configure connect as the entrypoint
ENTRYPOINT ["/usr/bin/connect-client"]
```

```

For example, to run a basic Python-based HTTP server, build a container from the following Dockerfile. When a data studio runs the custom template environment, the value for the `CONNECT_TOOL_PORT` environment variable is provided dynamically.

```docker title="Example Dockerfile with Python HTTP server"
FROM public.cr.seqera.io/platform/connect-client:0.7 as connect

FROM ubuntu:20.04
RUN apt-get update --yes && apt-get install --yes --no-install-recommends python3

COPY --from=connect /connect-client /usr/bin/connect-client

RUN apt-get update --yes && apt-get install --yes --no-install-recommends \
    ca-certificates \
    wget \
    fuse \
    btrfs-progs \
    squashfs-tools \
    && apt-get clean && rm -rf /var/lib/apt/lists/*

RUN wget https://fusionfs.seqera.io/releases/pkg/2/4/2/fusion-amd64.tar.gz && \
    tar xvzf fusion-amd64.tar.gz && \
    rm fusion-amd64.tar.gz

ENV FUSION_LOG_OUTPUT=/tmp/.fusion.log
ENV FUSION_COMMAND_BUFFER_SIZE=0
ENV FUSION_VALIDATE_WORK=false

ENTRYPOINT ["/usr/bin/fusion", "/usr/bin/connect-client"]

CMD ["/usr/bin/bash", "-c", "python3 -m http.server $CONNECT_TOOL_PORT"]

> 

Reviewer 2: Updated example:
```
FROM public.cr.seqera.io/platform/connect-client:0.7 as connect

FROM ubuntu:20.04
RUN apt-get update --yes && apt-get install --yes --no-install-recommends python3

COPY --from=connect /connect-client /usr/bin/connect-client
RUN /usr/bin/connect-client --install
ENTRYPOINT ["/usr/bin/connect-client"]

CMD ["/usr/bin/bash", "-c", "python3 -m http.server $CONNECT_TOOL_PORT"]
```

```

> ## Inspect custom container template build status {#build-status}

Reviewer 2: This section only applies to augmenting the Seqera-provided images but not to custom container template images. It might just be me, but I find the wording (Inspect custom container template) suggests that it applies to custom container template images as well

Author: @jimmypoms, ah, okay. I thought it was both. I can move this under the augmentation section and change the title.
